File: cpp98/Ice/invoke/PrinterI.cpp
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

#include <Ice/Ice.h>
#include <PrinterI.h>

using namespace std;

static ostream&
operator<<(ostream& out, Demo::Color c)
{
    switch(c)
    {
    case Demo::red:
        out << "red";
        break;
    case Demo::green:
        out << "green";
        break;
    case Demo::blue:
        out << "blue";
        break;
    }
    return out;
}

bool
PrinterI::ice_invoke(const vector<Ice::Byte>& inParams, vector<Ice::Byte>& outParams, const Ice::Current& current)
{
    Ice::CommunicatorPtr communicator = current.adapter->getCommunicator();

    bool result = true;

    Ice::InputStream in(communicator, inParams);
    in.startEncapsulation();

    if(current.operation == "printString")
    {
        string message;
        in.read(message);
        cout << "Printing string `" << message << "'" << endl;
    }
    else if(current.operation == "printStringSequence")
    {
        Demo::StringSeq seq;
        in.read(seq);
        cout << "Printing string sequence {";
        for(Demo::StringSeq::iterator p = seq.begin(); p != seq.end(); ++p)
        {
            if(p != seq.begin())
            {
                cout << ", ";
            }
            cout << "'" << *p << "'";
        }
        cout << "}" << endl;
    }
    else if(current.operation == "printDictionary")
    {
        Demo::StringDict dict;
        in.read(dict);
        cout << "Printing dictionary {";
        for(Demo::StringDict::iterator p = dict.begin(); p != dict.end(); ++p)
        {
            if(p != dict.begin())
            {
                cout << ", ";
            }
            cout << p->first << "=" << p->second;
        }
        cout << "}" << endl;
    }
    else if(current.operation == "printEnum")
    {
        Demo::Color c;
        in.read(c);
        cout << "Printing enum " << c << endl;
    }
    else if(current.operation == "printStruct")
    {
        Demo::Structure s;
        in.read(s);
        cout << "Printing struct: name=" << s.name << ", value=" << s.value << endl;
    }
    else if(current.operation == "printStructSequence")
    {
        Demo::StructureSeq seq;
        in.read(seq);
        cout << "Printing struct sequence: {";
        for(Demo::StructureSeq::iterator p = seq.begin(); p != seq.end(); ++p)
        {
            if(p != seq.begin())
            {
                cout << ", ";
            }
            cout << p->name << "=" << p->value;
        }
        cout << "}" << endl;
    }
    else if(current.operation == "printClass")
    {
        Demo::CPtr c;
        in.read(c);
        cout << "Printing class: s.name=" << c->s.name << ", s.value=" << c->s.value << endl;
    }
    else if(current.operation == "getValues")
    {
        Demo::CPtr c = new Demo::C;
        c->s.name = "green";
        c->s.value = Demo::green;
        Ice::OutputStream out(communicator);
        out.startEncapsulation();
        out.write(c);
        out.write("hello");
        out.endEncapsulation();
        out.finished(outParams);
    }
    else if(current.operation == "throwPrintFailure")
    {
        cout << "Throwing PrintFailure" << endl;
        Demo::PrintFailure ex;
        ex.reason = "paper tray empty";
        Ice::OutputStream out(communicator);
        out.startEncapsulation();
        out.write(ex);
        out.endEncapsulation();
        out.finished(outParams);
        result = false;
    }
    else if(current.operation == "shutdown")
    {
        current.adapter->getCommunicator()->shutdown();
    }
    else
    {
        Ice::OperationNotExistException ex(__FILE__, __LINE__);
        ex.id = current.id;
        ex.facet = current.facet;
        ex.operation = current.operation;
        throw ex;
    }

    in.endEncapsulation();
    return result;
}

File: swift/Ice/invoke/Sources/Server/PrinterI.swift
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import Foundation
import Ice

extension Structure: CustomStringConvertible {
    public var description: String {
        return "name=\(name), value=\(value)"
    }
}

extension C: CustomStringConvertible {
    public var description: String {
        return "s.name=\(s.name), s.value=\(s.value)"
    }
}

class PrinterI: Ice.Blobject {
    func ice_invoke(inEncaps: Data, current: Current) throws -> (ok: Bool, outParams: Data) {
        let communicator = current.adapter!.getCommunicator()

        precondition(inEncaps.count >= 6, "Invalid incapsulation")
        let inStream = Ice.InputStream(communicator: communicator, bytes: inEncaps)
        try inStream.startEncapsulation()

        var ret: (ok: Bool, outParams: Data)
        switch current.operation {
        case "printString":
            let message: String = try inStream.read()
            print("Printing string `\(message)'")
            ret = (ok: true, outParams: Data())
        case "printStringSequence":
            let seq: StringSeq = try inStream.read()
            print("Printing string sequence \(seq)")
            ret = (ok: true, outParams: Data())
        case "printDictionary":
            let dict = try StringDictHelper.read(from: inStream)
            print("Printing dictionary \(dict)")
            ret = (ok: true, outParams: Data())
        case "printEnum":
            let c: Color = try inStream.read()
            print("Printing enum \(c)")
            ret = (ok: true, outParams: Data())
        case "printStruct":
            let s: Structure = try inStream.read()
            print("Printing struct: \(s)")
            ret = (ok: true, outParams: Data())
        case "printStructSequence":
            let seq = try StructureSeqHelper.read(from: inStream)
            print("Printing struct sequence: \(seq)")
            ret = (ok: true, outParams: Data())
        case "printClass":
            var c: C?
            try inStream.read { c = $0 as? C }
            try inStream.readPendingValues()
            try inStream.endEncapsulation()
            print("Printing class: \(c!)")
            return (ok: true, outParams: Data())
        case "getValues":
            let c = C(s: Structure(name: "green", value: .green))
            let outStream = Ice.OutputStream(communicator: communicator)
            outStream.startEncapsulation()
            outStream.write(c)
            outStream.write("hello")
            outStream.writePendingValues()
            outStream.endEncapsulation()
            ret = (ok: true, outParams: outStream.finished())
        case "throwPrintFailure":
            print("Throwing PrintFailure")
            let ex = PrintFailure(reason: "paper tray empty")
            let outStream = Ice.OutputStream(communicator: communicator)
            outStream.startEncapsulation()
            outStream.write(ex)
            outStream.endEncapsulation()
            ret = (ok: false, outParams: outStream.finished())
        case "shutdown":
            current.adapter!.getCommunicator().shutdown()
            ret = (ok: true, outParams: Data())
        default:
            throw Ice.OperationNotExistException(id: current.id, facet: current.facet, operation: current.operation)
        }

        try inStream.endEncapsulation()
        return ret
    }
}
